x-pack/plugins/siem/server/lib/detection_engine/signals/search_after_bulk_create.ts

```diff
@@ -152,25 +116,62 @@ export const searchAfterAndBulkCreate = async ({
         pageSize, // maximum number of docs to receive per search result.
       });
       toReturn.searchAfterTimes.push(searchDuration);
+      toReturn.lastLookBackDate =
+        searchResult.hits.hits.length > 0
+          ? new Date(
+              searchResult.hits.hits[searchResult.hits.hits.length - 1]?._source['@timestamp']
+            )
+          : null;
+      const totalHits =
+        typeof searchResult.hits.total === 'number'
+          ? searchResult.hits.total
+          : searchResult.hits.total.value;
+      logger.debug(`totalHits: ${totalHits}`);
+
+      // re-calculate maxResults to ensure if our search results
+      // are less than max signals, we are not attempting to
+      // create more signals than there are total search results.
+      maxResults = Math.min(totalHits, ruleParams.maxSignals);
+      searchResultSize += searchResult.hits.hits.length;
       if (searchResult.hits.hits.length === 0) {
         toReturn.success = true;
         return toReturn;
       }
-      hitsSize += searchResult.hits.hits.length;
-      logger.debug(`size adjusted: ${hitsSize}`);
-      sortIds = searchResult.hits.hits[0].sort;
-      if (sortIds == null) {
-        logger.debug('sortIds was empty on search');
+
+      // filter out the search results that match with the values found in the list.
+      // the resulting set are valid signals that are not on the allowlist.
+      let filteredEvents;
```

Comment:
I would do this so you don't have a starting point of `any` where it will then be narrowed but rather you have your type right off at the starting point and then you can narrow it away from the undefined state if you want to:

```ts
let filteredEvents: SignalSearchResponse | undefined;
```

You can also optionally just pull this into a function to remove the possibility of `undefined` or you can use a ternary like so:

```ts
      // filter out the search results that match with the values found in the list.
      // the resulting set are valid signals that are not on the allowlist.
      const filteredEvents =
        listClient != null
          ? await filterEventsAgainstList({
              listClient,
              exceptionsList,
              logger,
              eventSearchResult: searchResult,
            })
          : searchResult;

      if (filteredEvents.hits.hits.length === 0) {
        // everything in the events were allowed, so no need to generate signals
        toReturn.success = true;
        return toReturn;
      }
```

Then you don't have to add extra `!= null` checks for typescript to be happy.